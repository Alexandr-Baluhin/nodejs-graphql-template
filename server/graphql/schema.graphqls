# Root _Query_ definition
type Query {
  # Get _Jokes_
  jokes: Jokes
}

# _Jokes_ definition
type Jokes {
    # Get _Jokes_ by _Category_
    byCategory(category: JokeCategoryEnum): Joke
}

# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available Joke's categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}
