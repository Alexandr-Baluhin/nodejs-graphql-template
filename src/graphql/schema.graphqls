# Root _Query_ definition
type Query {
  # Get _Jokes_
  jokes: Jokes
}

# _Jokes_ definition
type Jokes {
    # Get _Joke_ by _JokeCategoryEnum_
    byCategory(category: JokeCategoryEnum): Joke
}

# _Joke_ definition
type Joke {
  id: String!
  text: String!
  categories: [JokeCategoryEnum]
}

# Available _Joke_ categories
enum JokeCategoryEnum {
  NERDY
  EXPLICIT
}
