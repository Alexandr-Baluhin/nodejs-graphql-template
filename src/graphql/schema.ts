import fs from 'fs';
import path from 'path';
import { makeExecutableSchema } from 'graphql-tools';

import resolvers from './resolvers';

const typeDefs: string = fs.readFileSync(path.join(__dirname, './schema.graphql'), 'utf8');

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

export default schema;
