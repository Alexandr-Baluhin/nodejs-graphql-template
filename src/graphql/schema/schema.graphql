# import Query.* from "jokes.graphql"
