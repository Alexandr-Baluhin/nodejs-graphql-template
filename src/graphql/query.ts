import {
  GraphQLObjectType,
} from 'graphql';
import JokeCategoryEnum from './types/joke-category';
import Joke from './types/joke';
import jokeResolver from './resolvers/joke';

const Query = new GraphQLObjectType({
  name: 'Query',
  fields: () => ({
    joke: {
      args: {
        category: {
          type: JokeCategoryEnum,
          defaultValue: 'nerdy'
        }
      },
      type: Joke,
      resolve: jokeResolver
    },
  })
});

export default Query;
