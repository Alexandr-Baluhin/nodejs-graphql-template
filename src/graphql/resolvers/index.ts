import { merge } from 'lodash';
import { Resolvers } from '../_generated/types';

import { query as Query } from './Query';
import { jokes as Jokes } from './Jokes';
import { joke as Joke } from './Joke';

const resolvers: Resolvers = {
  Query,
  Jokes,
  Joke,
};

export default merge(resolvers);
