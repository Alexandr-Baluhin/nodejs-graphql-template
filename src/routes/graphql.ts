import * as graphqlHTTP from 'koa-graphql';
import schema from '../graphql/schema';

export default graphqlHTTP({
  schema,
  graphiql: process.env.GRAPHIQL
});
