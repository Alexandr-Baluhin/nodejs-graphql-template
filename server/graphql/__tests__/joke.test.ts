import { graphql } from 'graphql';
import schema from '../schema';
import * as nock from 'nock';

const rootValue = {};
const context = {};

beforeAll(() => {

  // mock service endpoint

  const defaultResponse = {
    type: 'success',
    value: {
      id: 2000,
      joke: `Chuck Norris can write to an output stream.`,
      categories: [],
    },
  };
  nock(process.env.JOKE_SERVICE_URI!)
    .get('/jokes/random')
    .reply(200, defaultResponse);

  const nerdyResponse = {
    type: 'success',
    value: {
      id: 1000,
      joke: `Mock Chuck Norris's database has only one table, 'Kick', which he DROPs frequently.`,
      categories: [
        'nerdy',
      ],
    },
  };
  nock(process.env.JOKE_SERVICE_URI!)
    .get('/jokes/random')
    .query({ limitTo: '[nerdy]' })
    .reply(200, nerdyResponse);
});

describe('query.joke', () => {
  it('should match default snapshot', async () => {
    const query = `
      query Q {
        jokeByCategory {
          id
          text
          categories
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context);
    const { data } = result;

    expect(data).toMatchSnapshot();
  });

  it('should match nerdy snapshot', async () => {
    const query = `
      query Q {
        jokeByCategory(category: NERDY) {
          id
          text
          categories
        }
      }
    `;

    const result = await graphql(schema, query, rootValue, context);
    const { data } = result;

    expect(data).toMatchSnapshot();
  });
});

