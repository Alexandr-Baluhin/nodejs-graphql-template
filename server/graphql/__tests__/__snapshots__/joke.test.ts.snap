// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query.joke should match default snapshot 1`] = `
Object {
  "data": Object {
    "jokeByCategory": Object {
      "categories": Array [],
      "id": "2000",
      "text": "Chuck Norris can write to an output stream.",
    },
  },
  "errors": undefined,
}
`;

exports[`query.joke should match nerdy snapshot 1`] = `
Object {
  "data": Object {
    "jokeByCategory": Object {
      "categories": Array [
        "NERDY",
      ],
      "id": "1000",
      "text": "Mock Chuck Norris's database has only one table, 'Kick', which he DROPs frequently.",
    },
  },
  "errors": undefined,
}
`;
